sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/epkversion/LeaseStoreManagerImpl.java

```diff
@@ -205,6 +211,47 @@ public Mono<Void> delete(Lease lease) {
             .then();
     }
 
+    @Override
+    public Mono<Void> deleteAll(List<Lease> leases) {
+        checkNotNull(leases, "Argument 'leases' can not be null");
+
+        logger.info("Deleting all leases");
+        Map<String, CosmosItemIdentity> cosmosIdentityMap = new HashMap<>();
+        for (Lease lease : leases) {
+            cosmosIdentityMap.put(lease.getId(), new CosmosItemIdentity(new PartitionKey(lease.getId()), lease.getId()));
+        }
+
+        return Mono.defer(() -> Mono.just(cosmosIdentityMap))
+            .flatMapMany(itemIdentities ->
+                this.leaseDocumentClient.deleteAllItems(cosmosIdentityMap.values().stream().collect(Collectors.toList())))
+            .flatMap(itemResponse -> {
+                if (itemResponse.getResponse() != null && itemResponse.getResponse().isSuccessStatusCode()) {
+                    cosmosIdentityMap.remove(itemResponse.getOperation().getId());
+                } else {
+                    // should ignore 404/0 for delete, will retry on other cases
+                    int effectiveStatusCode = 0;
+                    int effectiveSubStatusCode = 0;
+                    if (itemResponse.getResponse() != null) {
+                        effectiveStatusCode = itemResponse.getResponse().getStatusCode();
+                        effectiveSubStatusCode = itemResponse.getResponse().getStatusCode();
+                    } else if (itemResponse.getException() != null && itemResponse.getException() instanceof CosmosException) {
+                        CosmosException cosmosException = (CosmosException) itemResponse.getException();
+                        effectiveStatusCode = cosmosException.getStatusCode();
+                        effectiveSubStatusCode = cosmosException.getSubStatusCode();
+                    }
+
+                    if (effectiveStatusCode == ChangeFeedHelper.HTTP_STATUS_CODE_NOT_FOUND &&
+                        effectiveSubStatusCode == 0) {
+                        cosmosIdentityMap.remove(itemResponse.getOperation().getId());
+                    }
+                }
+
+                return Mono.empty();
+            })
+            .repeat(() -> cosmosIdentityMap.size() != 0)
```

Comment:
should I quite after 10 times instead of forever?

Reply:
No - repetition is the right choice here - allows early identification if there is an issue.